### PR TITLE
Integrate the C++ based gRPC batcher into Mukoe

### PR DESCRIPTION
**Pull request by allenwang28**
We can demonstrate a slight improvement with TPUs!

Mukoe e2e on 1x repr TPU:

```
([MuZeroActor: 25](muzero-cluster-kuberay-worker-cpuworker-krghm) pid=916, ip=10.116.96.49) INFO:absl:Note: It takes 0.35530519485473633 to update actor
([MuZeroActor: 25](muzero-cluster-kuberay-worker-cpuworker-krghm) pid=916, ip=10.116.96.49) INFO:absl:{'act/episode_length': 811, 'act/episode_return': -21.0, 'act/episode_duration': 371.184024810791, 'act/steps_per_second': 2.1849000651722625, 'act/select_action_latency_sec': 0.4537861749952436, 'act/env_reset_latency_sec': 0.011235952377319336, 'act/env_step_latency_sec': 0.002258673549139397, 'act/acting_network_step': 1000, 'episodes': 1, 'steps': 811, 'mcts_search_height_mean': 3.558471023427867, 'mcts_search_height_max': 11.998766954377311, 'raw_value': -2.0289314, 'mcts_sampling_temperature': 1.0}
```

```
(RayInferenceActor pid=347, ip=10.116.107.30) MukoeBatcher statistics:
(RayInferenceActor pid=347, ip=10.116.107.30) -----------------------
(RayInferenceActor pid=347, ip=10.116.107.30) Total data points: 9255
(RayInferenceActor pid=347, ip=10.116.107.30) Average batch size: 2.69897
(RayInferenceActor pid=347, ip=10.116.107.30) Average end to end time (ms): 44.7777
(RayInferenceActor pid=347, ip=10.116.107.30) Average batch build time (ms): 5.10384
(RayInferenceActor pid=347, ip=10.116.107.30) Average Python conversion time (ms): 0
(RayInferenceActor pid=347, ip=10.116.107.30) Average process time (ms): 39.4346
(RayInferenceActor pid=347, ip=10.116.107.30) Average C++ conversion time (ms): 0
(RayInferenceActor pid=347, ip=10.116.107.30) Average postprocess time (ms): 0
```

Mukoe e2e on CPU:

```
([MuZeroActor: 10](muzero-cluster-kuberay-worker-cpuworker-gfrbp) pid=763, ip=10.116.52.47) INFO:absl:{'act/episode_length': 821, 'act/episode_return': -21.0, 'act/episode_duration': 402.9942283630371, 'act/steps_per_second': 2.0372500205149406, 'act/select_action_latency_sec': 0.486619931626407, 'act/env_reset_latency_sec': 0.014180183410644531, 'act/env_step_latency_sec': 0.002327531031656207, 'act/acting_network_step': 1000, 'episodes': 1, 'steps': 821, 'mcts_search_height_mean': 3.558538367844093, 'mcts_search_height_max': 12.0, 'raw_value': -2.0312424, 'mcts_sampling_temperature': 1.0}
```

### CODE REVIEW

**Review by WendyShang — APPROVED** [review]
AWESOME!!!! 
We got it yay!!!

File: muzero/ray_main.py [hunk 1]
```diff
@@ -66,50 +72,87 @@ def start_learner(self):
...
+            logging.info(
+                "repr_infer was set to tpu, creating %d TPU actors.", num_actors
+            )
+            for _ in range(num_actors):
+                self.repr_tpu_inferers.append(
+                    RayInferenceActor.options(
+                        num_cpus=1, resources={"repr_cpu_handler": 1}
+                    ).remote(
+                        ckpt_dir=args.ckpt_dir,
+                        batch_size=inference_config.dyna_batch_size,
```
Comment (WendyShang) [inline muzero/ray_main.py:117]: this should be repr batch size and repr time out

**Review by WendyShang — CHANGES_REQUESTED** [review]

File: muzero/ray_main.py [hunk 2]
```diff
@@ -66,50 +72,87 @@ def start_learner(self):
...
-                model="repr",
-                tpu_id=TPU_ID_REPR,
-                weight_update_interval=inference_config.repr_update_interval,
+            logging.info(
+                "dyna_infer was set to tpu, creating %d TPU actors.", num_actors
+            )
+            for _ in range(num_actors):
+                self.dyna_tpu_inferers.append(
+                    RayInferenceActor.options(
+                        num_cpus=1, resources={"dyna_cpu_handler": 1}
```
[the remark this answers is not in the document]
Comment (WendyShang) [reply muzero/ray_main.py:87]: There is other code change required except for the setup of the ray cluster and resource allocation:
for actor_index in range(num_actors):
                self.repr_tpu_inferers.append(
                    RayInferenceActor.options(
                        num_cpus=1, resources={
                            CPU_ID_REPR + '_' + str(actor_index+1): 1}
                    ).remote(
                        ckpt_dir=args.ckpt_dir,
                        batch_size=inference_config.repr_batch_size,
                        batch_timeout_s=inference_config.repr_time_out,
                        model="repr",
                        tpu_id=TPU_ID_REPR + '_' + str(actor_index+1),
                        weight_update_interval=inference_config.repr_update_interval,
                    )
                )

[the remark this answers is not in the document]
Comment (allenwang28) [reply muzero/ray_main.py]: done

File: muzero/ray_main.py (hunk 1 above)
Comment (allenwang28) [reply muzero/ray_main.py]: done

**Review by WendyShang — APPROVED** [review]

File: muzero/ray_main.py [hunk 3]
```diff
@@ -66,50 +75,87 @@ def start_learner(self):
...
-            self.inference_actor_dyna = None
-            self.inference_actor_repr = None
-        else:
-            assert "wrong inference node type specified: cpu or tpu only"
 
     def start_muzero_actor(self, actor_i: int):
         args = self.args
-        if actor_i % 100 == 0:
+        # if actor_i % 100 == 0:
+        if actor_i % 1 == 0:
```
[the remark this answers is not in the document]
Comment (allenwang28) [reply muzero/ray_main.py]: fixed

**Pull request merged**